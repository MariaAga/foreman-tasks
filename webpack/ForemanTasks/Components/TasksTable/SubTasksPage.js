import React from 'react';
import PropTypes from 'prop-types';
import { translate as __ } from 'foremanReact/common/I18n';
import TasksTablePage from './';

export const SubTasksPage = props => {
  const parentTaskID = props.match.params.id;
  const getBreadcrumbs = actionName => ({
    breadcrumbItems: [
      { caption: __('Tasks'), url: `/foreman_tasks/tasks` },
      {
        caption: actionName,
        url: `/foreman_tasks/tasks/${parentTaskID}`,
      },
      { caption: __('Sub tasks') },
    ],
  });
  return (
    <TasksTablePage
      getBreadcrumbs={getBreadcrumbs}
      parentTaskID={parentTaskID}
      {...props}
    />
  );
};

SubTasksPage.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.object,
  }),
};

SubTasksPage.defaultProps = {
  match: {
    params: {},
  },
};
