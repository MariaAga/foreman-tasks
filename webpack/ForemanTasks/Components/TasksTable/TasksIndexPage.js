import React from 'react';
import { translate as __ } from 'foremanReact/common/I18n';
import TasksTablePage from './';

export const TasksIndexPage = props => {
  const getBreadcrumbs = () => ({
    breadcrumbItems: [{ caption: __('Tasks'), url: `/foreman_tasks/tasks` }],
  });
  return <TasksTablePage getBreadcrumbs={getBreadcrumbs} {...props} />;
};
