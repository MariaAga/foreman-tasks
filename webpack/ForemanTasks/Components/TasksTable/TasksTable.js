import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import { Table } from 'foremanReact/components/common/table';
import { STATUS } from 'foremanReact/constants';
import MessageBox from 'foremanReact/components/common/MessageBox';
import { translate as __ } from 'foremanReact/common/I18n';
import Pagination from 'foremanReact/components/Pagination/PaginationWrapper';
import { getURIQuery } from 'foremanReact/common/helpers';
import createTasksTableSchema from './TasksTableSchema';
import { updateURlQuery } from './TasksTableHelpers';

const TasksTable = ({
  getTableItems,
  error,
  status,
  results,
  history,
  itemCount,
  pagination,
}) => {
  const url = history.location.pathname + history.location.search;
  const uriQuery = getURIQuery(url);

  useEffect(() => {
    getTableItems(url);
  }, [history.location.search]);

  if (status === STATUS.ERROR) {
    return (
      <MessageBox
        key="tasks-table-error"
        icontype="error-circle-o"
        msg={__(`Could not receive data: ${error && error.message}`)}
      />
    );
  }

  if (status === STATUS.PENDING && results.length === 0) {
    return <div />;
  }

  if (results.length === 0) {
    return <span>{__('No Tasks')}</span>;
  }

  const changeTablePage = ({ page, perPage }) => {
    updateURlQuery(
      {
        page,
        per_page: perPage,
      },
      history
    );
  };

  const setSortHistory = (by, order) => {
    updateURlQuery({ sort_by: by, sort_order: order }, history);
  };

  return (
    <div className="tasks-table">
      <Table
        key="tasks-table"
        columns={createTasksTableSchema(
          setSortHistory,
          uriQuery.sort_by,
          uriQuery.sort_order
        )}
        rows={results}
      />
      <Pagination
        className="tasks-pagination"
        viewType="table"
        itemCount={itemCount}
        pagination={pagination}
        onChange={changeTablePage}
        dropdownButtonId="tasks-table-dropdown"
      />
    </div>
  );
};

TasksTable.propTypes = {
  results: PropTypes.array.isRequired,
  getTableItems: PropTypes.func.isRequired,
  status: PropTypes.oneOf(Object.keys(STATUS)),
  error: PropTypes.instanceOf(Error),
  itemCount: PropTypes.number.isRequired,
  pagination: PropTypes.shape({
    page: PropTypes.number,
    perPage: PropTypes.number,
  }),
  history: PropTypes.object.isRequired,
};

TasksTable.defaultProps = {
  status: STATUS.PENDING,
  error: null,
  pagination: {
    page: 1,
    perPage: 20,
  },
};

export default TasksTable;
