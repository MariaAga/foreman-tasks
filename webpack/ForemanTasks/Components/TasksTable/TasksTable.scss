.tasks-pagination {
  margin-top: -6px;
}

#tasks-table {
  margin-bottom: 70px;
}
