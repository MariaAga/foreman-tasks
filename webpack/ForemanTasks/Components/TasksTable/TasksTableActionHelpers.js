import { translate as __, sprintf } from 'foremanReact/common/I18n';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { TASKS_DASHBOARD_AVAILABLE_QUERY_MODES } from '../TasksDashboard/TasksDashboardConstants';
import { timeToHoursNumber } from '../TasksDashboard/TasksDashboardHelper';

export const convertDashboardQuery = query => {
  const {
    time_mode: timeMode,
    time_horizon: timeHorizon,
    state,
    result,
    search,
  } = query;

  const hours = timeToHoursNumber(timeHorizon);

  const timestamp = new Date(new Date() - hours * 60 * 60 * 1000);
  let dashboardTime = '';
  const stateQuery = state ? `state=${state}` : '';
  const resultQuery = result ? `result=${result}` : '';
  if (timeMode === TASKS_DASHBOARD_AVAILABLE_QUERY_MODES.LAST) {
    dashboardTime = `(state_updated_at<${timestamp.toISOString()} or state_updated_at = NULL)`;
  } else if (timeMode === TASKS_DASHBOARD_AVAILABLE_QUERY_MODES.OLDER) {
    dashboardTime = `(state_updated_at>${timestamp.toISOString()})`;
  }
  const newQuery = [stateQuery, resultQuery, search, dashboardTime]
    .filter(Boolean)
    .join(' and ');
  return newQuery;
};

export const resumeToastInfo = {
  resumed: { type: 'success', text: __('was resumed') },
  failed: { type: 'error', text: __('could not be resumed') },
  skipped: { type: 'warning', text: __('task has to be resumable') },
};

export const cancelToastInfo = {
  cancelled: { type: 'success', text: __('was cancelled') },
  skipped: { type: 'warning', text: __('task has to be cancellable') },
};

export const toastDispatch = ({ type, name, toastInfo, dispatch }) => {
  dispatch(
    addToast({
      type: toastInfo[type].type,
      message: sprintf('%(name)s Task execution %(type)s', {
        name,
        type: toastInfo[type].text,
      }),
    })
  );
};
