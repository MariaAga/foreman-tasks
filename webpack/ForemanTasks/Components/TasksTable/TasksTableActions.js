import { getURIQuery } from 'foremanReact/common/helpers';
import { getTableItemsAction } from 'foremanReact/components/common/table';
import API from 'foremanReact/API';
import { addToast } from 'foremanReact/redux/actions/toasts';
import { translate as __ } from 'foremanReact/common/I18n';
import { TASKS_TABLE_ID } from './TasksTableConstants';
import { getApiPathname } from './TasksTableHelpers';
import { fetchTasksSummary } from '../TasksDashboard/TasksDashboardActions';

export const getTableItems = url =>
  getTableItemsAction(TASKS_TABLE_ID, getURIQuery(url), getApiPathname(url));

export const cancelTaskAction = (id, name, url) => async dispatch => {
  await dispatch(cancelTask(id, name));
  dispatch(getTableItems(url));
  dispatch(fetchTasksSummary(getURIQuery(url).time));
};

export const cancelTask = (id, name) => async dispatch => {
  dispatch(
    addToast({
      type: 'info',
      message: `${__('Trying to cancel')} "${name}" ${__('task')}`,
    })
  );
  try {
    await API.post(`/foreman_tasks/tasks/${id}/cancel`);
    dispatch(
      addToast({
        type: 'success',
        message: `"${name}" ${__('task cancelled')}`,
      })
    );
  } catch ({ response }) {
    dispatch(
      addToast({
        type: 'error',
        message: `"${name}" ${__('task cannot be cancelled at the moment.')}`,
      })
    );
  }
};
