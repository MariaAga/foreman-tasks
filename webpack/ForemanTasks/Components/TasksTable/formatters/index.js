export { default as selectionHeaderCellFormatter } from './selectionHeaderCellFormatter';
export { default as selectionCellFormatter } from './selectionCellFormatter';

export { actionNameCellFormatter } from './actionNameCellFormatter';
export { actionCellFormatter } from './actionCellFormatter';
export { dateCellFormmatter } from './dateCellFormmatter';
export { durationCellFormmatter } from './durationCellFormmatter';
