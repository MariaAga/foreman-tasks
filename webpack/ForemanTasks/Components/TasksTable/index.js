import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import TasksTable from './TasksTable';
import reducer from './TasksTableReducer';
import * as actions from './TasksTableActions';
import { selectQuery } from '../TasksDashboard/TasksDashboardSelectors';
import {
  selectStatus,
  selectError,
  selectResults,
  selectPagitation,
  selectItemCount,
  selectSort,
} from './TasksTableSelectors';

const mapStateToProps = state => ({
  status: selectStatus(state),
  error: selectError(state),
  sort: selectSort(state),
  results: selectResults(state),
  query: selectQuery(state),
  pagination: selectPagitation(state),
  itemCount: selectItemCount(state),
});
const mapDispatchToProps = dispatch => bindActionCreators(actions, dispatch);

export const reducers = { tasksTable: reducer };

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(TasksTable);
