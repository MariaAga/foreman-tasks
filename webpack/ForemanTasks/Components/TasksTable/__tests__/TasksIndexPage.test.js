import { testComponentSnapshotsWithFixtures } from 'react-redux-test-utils';
import { TasksIndexPage } from '../TasksIndexPage';
import { minProps } from './TasksTable.fixtures';

const fixtures = {
  'render with minimal props': minProps,
};

describe('TasksIndexPage', () => {
  describe('rendering', () =>
    testComponentSnapshotsWithFixtures(TasksIndexPage, fixtures));
});
