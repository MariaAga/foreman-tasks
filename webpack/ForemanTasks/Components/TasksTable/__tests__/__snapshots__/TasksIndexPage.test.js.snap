// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TasksIndexPage rendering render with minimal props 1`] = `
<Connect(TasksTablePage)
  cancelTaskAction={[MockFunction]}
  getBreadcrumbs={[MockFunction]}
  getTableItems={[MockFunction]}
  history={
    Object {
      "location": Object {
        "search": "",
      },
    }
  }
  itemCount={2}
  location={
    Object {
      "pathname": "",
    }
  }
  pagination={
    Object {
      "page": 1,
      "perPage": 10,
    }
  }
  results={
    Array [
      "a",
      "b",
    ]
  }
  sort={
    Object {
      "by": "q",
      "order": "w",
    }
  }
  status="RESOLVED"
/>
`;
